library/src/main/java/com/parse/internal/signpost/basic/HttpURLConnectionProvider.java

```diff
@@ -0,0 +1,56 @@
+package com.parse.internal.signpost.basic;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+/**
+ * Singleton class to provide HttpURLConnection, using native HttpURLConnection or OkHttp
+ */
+public class HttpURLConnectionProvider {
+    private static final String TAG = "FacebookActivity";
+
+    boolean isUsingOkHttp = false;
+    Object okUrlFactory = null;
+    Method okUrlFactoryOpen = null;
+
+    protected HttpURLConnectionProvider() {
+        try {
+            Class okHttpClientClass = Class.forName("com.squareup.okhttp.OkHttpClient");
+            Object okHttpClient = okHttpClientClass.getConstructor().newInstance();
+            Class okUrlFactoryClass = Class.forName("com.squareup.okhttp.OkUrlFactory");
+            okUrlFactory = okUrlFactoryClass.getConstructor(okHttpClientClass).newInstance(okHttpClient);
+            okUrlFactoryOpen = okUrlFactoryClass.getMethod("open", URL.class);
+            isUsingOkHttp = true;
+        } catch (Exception e) {
+            isUsingOkHttp = false;
+        }
+    }
+
+    private static HttpURLConnectionProvider instance = null;
+    public static HttpURLConnectionProvider getInstance() {
+        if(instance == null) {
+            synchronized(HttpURLConnectionProvider.class) {
+                if(instance == null) {
+                    instance = new HttpURLConnectionProvider();
+                }
+            }
+        }
+        return instance;
+    }
+
+    public HttpURLConnection open(URL url) throws IOException {
+        HttpURLConnection connection;
+        if (isUsingOkHttp) {
+            try {
+                connection = (HttpURLConnection) okUrlFactoryOpen.invoke(okUrlFactory, url);
+            } catch (Exception e) {
+                connection = (HttpURLConnection) url.openConnection();
```

Comment:
OkUrlFactory.open itself doesn't throw, so exception is likely to be caused by OkHttp version incompatibility. In those cases, falling back to HttpURLConnection may be better.